### PR TITLE
BUG-22984 Fix truncation of DataFrame representations

### LINKED ISSUE #22984: BUG: wrong detection if truncated repr is needed
There seems to be a bug in the detection of when a DataFrame repr will not fit the console width.

Small example, the repr of a 7 column float dataframe is 72 characters wide:

```
In [10]: pd.DataFrame(np.random.randn(3, 7))
Out[10]: 
          0         1         2         3         4         5         6
0 -0.039631 -0.630607 -0.041318 -0.679839  0.257494 -0.608414  0.013959
1  0.976692  0.578279  0.712098  1.292541 -0.934307  0.522638 -0.584541
2 -1.218953  0.124633 -0.705128  0.376275  0.044354  0.166730  1.122929
```

However, in my default terminal of 80 characters wide, it is truncated:

```
In [1]: pd.io.formats.console.get_terminal_size()
Out[1]: os.terminal_size(columns=80, lines=24)

In [2]: pd.DataFrame(np.random.randn(3, 7))
Out[2]: 
          0         1         2    ...            4         5         6
0  0.332255  0.408991 -0.446875    ...     0.533822  0.128198  0.421326
1  0.335729 -0.889485  2.038889    ...    -0.263465  1.786649  1.190037
2  0.197376  1.742540 -1.259510    ...     0.643844  0.634065 -0.623578

[3 rows x 7 columns]
```

("funny" thing is that this truncated repr is actually exactly as wide as the non-truncated one ...)

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [X] closes #22984 
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

When printing a DataFrame to terminal, an extra column's worth of space is added to the calculated width of the DataFrame. This is presumably to help edge cases, but the calculated difference between the DataFrame width and the terminal window width is incremented by 1 a few lines later, seemingly to fix the same problem. Do any more experienced developers know of a reason to pad the DataFrame width even more?

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -194,6 +194,7 @@ Other Enhancements
 - :meth:`Index.to_frame` now supports overriding column name(s) (:issue:`22580`).
 - New attribute :attr:`__git_version__` will return git commit sha of current build (:issue:`21295`).
 - Compatibility with Matplotlib 3.0 (:issue:`22790`).
+- Representation of :class:`DataFrame` fills up the terminal window better
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you add the issue number here, I would call this more of a bug fix, no?

File: pandas/io/formats/format.py [hunk 2]
```diff
@@ -616,11 +616,6 @@ def to_string(self):
             else:  # max_cols == 0. Try to fit frame to terminal
                 text = self.adj.adjoin(1, *strcols).split('\n')
                 max_len = Series(text).str.len().max()
-                headers = [ele[0] for ele in strcols]
```
Comment (jreback) [inline pandas/io/formats/format.py]: we have a number of tests for this, suprised this didn't break anything, can you make a test if we have no coverage for this case now? (e.g. you use option_context to set the width, then check the output string)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
cc @jorisvandenbossche

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @TomAugspurger a glance if you can.

File: doc/source/whatsnew/v0.24.0.txt [hunk 3]
```diff
@@ -1312,6 +1312,7 @@ Notice how we now instead output ``np.nan`` itself instead of a stringified form
 - :func:`read_sas()` will correctly parse sas7bdat files with data page types having also bit 7 set (so page type is 128 + 256 = 384) (:issue:`16615`)
 - Bug in :meth:`detect_client_encoding` where potential ``IOError`` goes unhandled when importing in a mod_wsgi process due to restricted access to stdout. (:issue:`21552`)
 - Bug in :func:`to_string()` that broke column alignment when ``index=False`` and width of first column's values is greater than the width of first column's header (:issue:`16839`, :issue:`13032`)
+- Bug in :func:`to_string()` that caused representations of :class:`DataFrame` to not take up the whole window (:issue:`22984`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.txt]: This should be :class:\`DataFrame.to_string\` right? We don't have a top-level `pandas.to_string`.

File: pandas/tests/io/formats/test_format.py [hunk 4]
```diff
@@ -343,6 +343,16 @@ def test_repr_truncates_terminal_size(self):
 
         assert df2.columns[0] in result.split('\n')[0]
```
Comment (TomAugspurger) [inline pandas/tests/io/formats/test_format.py]: nitpick: we have a pytest fixture formock now, that does the try / except / skip done above.

I think it'd be cleanest to split this into a new test here, and accept the `mock` parameter.

```python
def test_repr_truncates_terminal_size_full(self, mock):
    ...
```

Any if you're feeling adventurous, you could change the try / except / skip mock import above to use the fixture as well. Not a big deal though.

**Comment by jreback** [comment]
@JustinZhengBC need to merge master

**Comment by TomAugspurger** [comment]
Another hypothesis failure on azure. OK to merge @jreback? I can investigate the failing test separately.

**Comment by jreback** [comment]
yep

**Comment by TomAugspurger** [comment]
Merged master to fix the confict. May as well let the CI run.

**Comment by JustinZhengBC** [comment]
@TomAugspurger green. Also thanks for fixing the merge issue

**Comment by TomAugspurger** [comment]
I caused it my merging my own PR, so it's the least I could do :)

**Pull request merged**